hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/MisReplicationHandler.java

```diff
@@ -0,0 +1,204 @@
+      Set<ContainerReplica> replicas, List<ContainerReplicaOp> pendingOps,
+      ContainerHealthResult result, int remainingMaintenanceRedundancy)
+      throws IOException {
+    ContainerInfo container = result.getContainerInfo();
+    ContainerReplicaCount replicaCount = getContainerReplicaCount(container,
+            replicas, pendingOps, remainingMaintenanceRedundancy);
+
+    if (!replicaCount.isSufficientlyReplicated() ||
+            replicaCount.isOverReplicated()) {
+      LOG.info("Container {} state should be neither under replicated " +
+              "nor over replicated before resolving misreplication." +
+              "Container UnderReplication status: {}," +
+              "Container OverReplication status: {}",
+              container.getContainerID(),
+              !replicaCount.isSufficientlyReplicated(),
+              replicaCount.isOverReplicated());
+      return Collections.emptyMap();
+    }
+    final Set<DatanodeDetails> deletionInFlight = new HashSet<>();
+    final Set<DatanodeDetails> addInFlight = new HashSet<>();
+
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        deletionInFlight.add(op.getTarget());
+      } else if (op.getOpType() == ContainerReplicaOp.PendingOpType.ADD) {
+        addInFlight.add(op.getTarget());
+      }
+    }
+
+    List<DatanodeDetails> usedDns = replicas.stream()
+            .map(ContainerReplica::getDatanodeDetails)
+            .collect(Collectors.toList());
+    if (containerPlacement.validateContainerPlacement(usedDns,
+            usedDns.size()).isPolicySatisfied()) {
+      LOG.info("Container {} is currently not misreplicated",
+              container.getContainerID());
+      return Collections.emptyMap();
+    }
+    List<DatanodeDetails> dnsWithPendingOps =
+            Stream.concat(usedDns.stream(), addInFlight.stream())
+            .filter(dn -> !deletionInFlight.contains(dn))
+            .distinct()
+            .collect(Collectors.toList());
+    if (containerPlacement.validateContainerPlacement(dnsWithPendingOps,
+            usedDns.size()).isPolicySatisfied()) {
+      LOG.info("Container {} is currently misreplicated currently" +
+               " but will not be misreplicated after pending ops executed",
+              container.getContainerID());
+      return Collections.emptyMap();
+    }
+    Set<ContainerReplica> sources = filterSources(replicas, deletionInFlight);
+    Set<ContainerReplica> replicasToBeReplicated = containerPlacement
+            .replicasToCopyToFixMisreplication(replicas.stream()
+            .collect(Collectors.toMap(Function.identity(), sources::contains)));
+    usedDns = replicas.stream().filter(r -> !replicasToBeReplicated.contains(r))
+            .map(ContainerReplica::getDatanodeDetails)
+            .collect(Collectors.toList());
+    List<DatanodeDetails> excludedDns = replicasToBeReplicated.stream()
```

Comment:
I think you also need to add pendingAdd to the exclude list.

But I am also a little concerned about the pending ops generally. We decided to keep `containerPlacement.replicasToCopyToFixMisreplication()` simple so it can only deal with "perfectly replicated". But if we can have a pendingAdd in flight, then it would affect what `replicasToCopyToFixMisreplication` should do. Perhaps its a pending add from a previous run to fix mis-replication, copying index 1 from rack 1. If we run the same lookup again, we may get index 1 to copy again, and that doesn't help.

Maybe we could set anything in pending add as "cannot copy" so we don't get another copy of it.

But then, what about pending delete. If the container is not over or under replicated, and there is somehow a pending delete, then its going to become under replicated.

I wonder if we would be better off not handling the mis-rep is there is any pending ops. Let the pending complete and then let the next iteration check again. I am not sure about this, but it would surely make it simpler.

Reply:
Removed the pending ops part.